node/src/accountant/payment_adjuster/mod.rs

```diff
@@ -0,0 +1,2340 @@
+                }
+
+                self.adjust_remaining_unallocated_cw_balance_down(&decided_accounts);
+                decided_accounts
+            }
+        };
+
+        let down_stream_decided_accounts = self.propose_adjustments_recursively(
+            remaining_undecided_accounts,
+            ServiceFeeOnlyAdjustmentRunner {},
+        );
+
+        RecursionResults::new(here_decided_accounts, down_stream_decided_accounts)
+    }
+
+    fn calculate_weights(&self, accounts: Vec<AnalyzedPayableAccount>) -> Vec<WeightedPayable> {
+        self.apply_criteria(self.calculators.as_slice(), accounts)
+    }
+
+    fn apply_criteria(
+        &self,
+        criteria_calculators: &[Box<dyn CriterionCalculator>],
+        qualified_accounts: Vec<AnalyzedPayableAccount>,
+    ) -> Vec<WeightedPayable> {
+        qualified_accounts
+            .into_iter()
+            .map(|payable| {
+                let weight =
+                    criteria_calculators
+                        .iter()
+                        .fold(0_u128, |weight, criterion_calculator| {
+                            let new_criterion = criterion_calculator
+                                .calculate(&payable.qualified_as, self.inner.as_ref());
+
+                            let summed_up = weight + new_criterion;
+
+                            calculated_criterion_and_weight_diagnostics(
+                                &payable.qualified_as.bare_account.wallet,
+                                criterion_calculator.as_ref(),
+                                new_criterion,
+                                summed_up,
+                            );
+
+                            summed_up
+                        });
+
+                WeightedPayable::new(payable, weight)
+            })
+            .collect()
+    }
+
+    fn adjust_remaining_unallocated_cw_balance_down(
+        &mut self,
+        processed_outweighed: &[AdjustedAccountBeforeFinalization],
+    ) {
+        let subtrahend_total: u128 = sum_as(processed_outweighed, |account| {
+            account.proposed_adjusted_balance_minor
+        });
+        self.inner
+            .subtract_from_unallocated_cw_service_fee_balance_minor(subtrahend_total);
+
+        diagnostics!(
+            "LOWERED CW BALANCE",
+            "Unallocated balance lowered by {} to {}",
+            subtrahend_total.separate_with_commas(),
+            self.inner
+                .unallocated_cw_service_fee_balance_minor()
+                .separate_with_commas()
+        )
+    }
+
+    fn sketch_debug_info_opt(
+        &self,
+        qualified_payables: &[AnalyzedPayableAccount],
+    ) -> Option<HashMap<Wallet, u128>> {
+        self.logger.debug_enabled().then(|| {
+            qualified_payables
+                .iter()
+                .map(|payable| {
+                    (
+                        payable.qualified_as.bare_account.wallet.clone(),
```

Comment:
I hope you'll use the `Address` here in case of the `Wallet`, it'll prevent `clone()`.